blocks/filesystem/base.py

```diff
@@ -1,18 +1,45 @@
-from abc import ABC, abstractmethod
+from typing import Union, Sequence, Tuple, List
+from collections import defaultdict
+from fsspec.core import split_protocol, get_filesystem_class, has_magic
 
-from typing import Union, IO, Sequence, Iterator, List, Any, Generator, ContextManager
-from blocks.datafile import DataFile
-from contextlib import contextmanager
 
+class FileSystem:
+    """Filesystem for manipulating files in the cloud
 
-class FileSystem(ABC):
-    """The required interface for any filesystem implementation
+    This supports operations on local files and any other protocol supported by fsspec.
+    This is a wrapper to fsspec which provides backwards compatibility for blocks filesystems
+    and a simplified interface.
 
-    See GCSFileSystem for a full implementation. This FileSystem is intended
-    to be extendable to support cloud file systems, encryption strategies, etc...
+    Parameters
+    ----------
+    storage_options: Mapping[str, Mapping[str, Any]]
+        Additional options passed to each filesystem for each protocol
+        e.g. {'gs': {'project': 'example'}} to set the gs filesytem project to example
     """
 
-    @abstractmethod
+    def __init__(self, **storage_options):
+        self.storage_options = defaultdict(dict)
+        self.storage_options.update(storage_options)
+        self.storage_options[None]["auto_mkdir"] = True
+        self.filesystems = {}
```

Comment:
nit: would be nice to instantiate the filesystem at this time with the given storage options, if present